### PR TITLE
feat(metadata): Add a generic way to add  metadata to different objects

### LINKED ISSUE #1610: Add meta data fields to core entities
# Overview 

Larger teams often require the ability to add their own meta-data to the following entities:

- Flags
- Segments
- Environments

Examples of the meta data that they might add:

- JIRA ticket code (e.g. `FLAGSMITH-152`) to a Flag
- URL for Infra overview for an Environment
- Link to Amplitude for a Segment

These data could be optional or required. For example, engineers should be required to add a JIRA ticket against a new Flag when the flag is created

# Implementation

Administrators should be able to define fields that they want associated with an entity, and whether it is required or optional.

To begin with the administrator should be able to define with the field is single line or multi-line. For v1, just allow any text. 

gz#40

### PR DESCRIPTION
**Pull request by gagantrivedi**
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/deployment/locally-api#pre-commit) to check linting
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?

## Changes

Add, a metadata app that can be used to add metadata to different objects
As a working proof of that, it adds the metadata to environment 

### New APIs
`​/api​/v1​/metadata​/fields​/`
`​/api​/v1​/metadata​/fields​/{ID}`
`​/api​/v1​/organisations​/{organisation_pk}​/metadata-model-fields​/`
`​/api​/v1​/organisations​/{organisation_pk}​/metadata-model-fields​/{ID}`
`​/api​/v1​/organisations​/{organisation_pk}​/metadata-model-fields​/supported-content-types​/`
`​/api​/v1​/organisations​/{organisation_pk}​/metadata-model-fields​/supported-required-for-models​/`

## How did you test this code?

Covered by unit test cases

### CODE REVIEW

File: api/metadata/models.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    is_required = models.BooleanField(default=False)
+
+    class Meta:
+        unique_together = ("field", "content_type")
+
+
+class Metadata(AbstractBaseExportableModel):
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    object_id = models.PositiveIntegerField()
+    content_object = GenericForeignKey("content_type", "object_id")
```
Comment (matthewelwell) [inline api/metadata/models.py:117]: Oh man, I didn't know this existed - that's super cool!

File: api/metadata/models.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+
+class FieldType(models.TextChoices):
+    INTEGER = "int"
+    STRING = "str"
+    BOOLEAN = "bool"
+    FLOAT = "float"
+
+
+class MetadataField(AbstractBaseExportableModel):
```
Comment (matthewelwell) [inline api/metadata/models.py:36]: So, as I understand it: 

This is a class for creating generic fields in an organisation?

...

File: api/metadata/models.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    name = models.CharField(max_length=255)
+    type = models.CharField(
+        max_length=255, choices=FieldType.choices, default=FieldType.STRING
+    )
+    description = models.TextField(blank=True, null=True)
+    organisation = models.ForeignKey(Organisation, on_delete=models.CASCADE)
+
+
+class MetadataModelField(AbstractBaseExportableModel):
```
Comment (matthewelwell) [inline api/metadata/models.py:80]: ... this is a class for linking a metadata field to another model class and specifying if it's required or not? 

...

File: api/metadata/models.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    field = models.ForeignKey(MetadataField, on_delete=models.CASCADE)
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    is_required = models.BooleanField(default=False)
+
+    class Meta:
+        unique_together = ("field", "content_type")
+
+
+class Metadata(AbstractBaseExportableModel):
```
Comment (matthewelwell) [inline api/metadata/models.py:109]: ... and this is a class which actually contains the data for a specific field against a specific model object? 

...

File: api/metadata/models.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
```
Comment (matthewelwell) [inline api/metadata/models.py:115]: ... what I don't understand is why we need content_type here when it's also defined against the MetadataModelField class (which we also link to below)?

File: api/metadata/models.py (hunk 2 above)
Comment (gagantrivedi) [reply api/metadata/models.py:36]: Yep

File: api/metadata/models.py (hunk 3 above)
Comment (gagantrivedi) [reply api/metadata/models.py:80]: Yep

File: api/metadata/models.py (hunk 4 above)
Comment (gagantrivedi) [reply api/metadata/models.py:109]: Yep

File: api/metadata/models.py (hunk 5 above)
Comment (gagantrivedi) [reply api/metadata/models.py:115]: Yeah, it's for creating the `GenericForeignKey` which is used for creating a reverse relation here:
https://github.com/Flagsmith/flagsmith/pull/1812/files#diff-a0b057e41b454fcf5879b6c8ddbb1cb10a4c446e671ef37ff3b242efa1e7200bR84

File: api/environments/serializers.py [hunk 6]
```diff
@@ -29,7 +30,9 @@ class Meta:
         )
 
 
-class EnvironmentSerializerLight(serializers.ModelSerializer):
+class EnvironmentSerializerLight(MetadataSerializerMixin, serializers.ModelSerializer):
```
Comment (matthewelwell) [inline api/environments/serializers.py]: Hmm... the point of the 'Light' serializer is to keep it light (i.e. not have to make extra queries, etc.). We shouldn't add the metadata here, we should perhaps create a totally new serializer.

File: api/environments/models.py [hunk 7]
```diff
@@ -91,6 +93,7 @@ class Environment(
     banner_colour = models.CharField(
         null=True, blank=True, max_length=7, help_text="hex code for the banner colour"
     )
+    metadata = GenericRelation(Metadata)
```
Comment (matthewelwell) [inline api/environments/models.py:101]: Man, this stuff is so cool!

File: api/import_export/export.py [hunk 8]
```diff
@@ -83,6 +85,18 @@ def export_organisation(organisation_id: int) -> typing.List[dict]:
...
 
 
+def export_metadata(organisation_id: int) -> typing.List[dict]:
+    return _export_entities(
+        _EntityExportConfig(MetadataField, Q(organisation__id=organisation_id)),
+        _EntityExportConfig(
+            MetadataModelField, Q(field__organisation__id=organisation_id)
+        ),
+        _EntityExportConfig(
+            Metadata, Q(model_field__field__organisation__id=organisation_id)
```
Comment (matthewelwell) [inline api/import_export/export.py:104]: Will this work without using the UUID to handle the relationship?

File: api/metadata/mixins.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+from .models import MetadataModelField
+from .serializers import MetaDataModelFieldSerializer
+
+
+class MetaDataViewMixin:
+    app_label = None
+    model_name = None
+
+    def get_base_metadata_fields_queryset(self):
```
Comment (matthewelwell) [inline api/metadata/mixins.py]: I think this method could be marked as 'private', right?

File: api/metadata/mixins.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+    @swagger_auto_schema(method="GET", responses={200: MetaDataModelFieldSerializer()})
+    @action(
+        detail=False,
+        methods=["GET"],
+        url_path="metadata-fields",
+        serializer_class=MetaDataModelFieldSerializer,
+    )
+    def get_model_metadata_fields(self, request):
+        content_type = ContentType.objects.get(
```
Comment (matthewelwell) [inline api/metadata/mixins.py]: Unless I'm yet to reach a point where app_label and model_name are used elsewhere, I think we can change this to `ContentType.objects.get_for_model(self.model_class)` and just define the model class against the view. Or perhaps we can even infer that from the queryset or something? ... but that's probably overkill.

File: api/metadata/mixins.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        responses={200: MetaDataModelFieldSerializer()},
+        request_body=MetaDataModelFieldSerializer(),
+    )
+    @action(
+        detail=False,
+        methods=["PUT"],
+        url_path=r"metadata-field/(?P<metadata_pk>\d+)",
+    )
+    def update_model_metadata_field(self, request, metadata_pk=None):
+        queryset = self.get_base_metadata_fields_queryset()
```
Comment (matthewelwell) [inline api/metadata/mixins.py]: Is there a reason we shouldn't filter based on the ContentType here? Just to save an extra query I guess?

File: api/metadata/mixins.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        method="POST",
+        responses={201: MetaDataModelFieldSerializer()},
+        request_body=MetaDataModelFieldSerializer(),
+    )
+    @action(detail=False, methods=["POST"], url_path="metadata-field")
+    def create_model_metadata_field(self, request):
+        serializer = MetaDataModelFieldSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        if not self.request.user.organisations.filter(
```
Comment (matthewelwell) [inline api/metadata/mixins.py]: Just playing devil's advocate here, is there a reason we wouldn't just create a viewset for MetadataModelFields nested under an organisation? The FE can then send up the content type as part of the payload?

File: api/metadata/mixins.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+        serializer = MetaDataModelFieldSerializer(
+            model_metadata_field, data=request.data, partial=True
+        )
+        serializer.is_valid(raise_exception=True)
+
+        serializer.save()
+        return Response(serializer.data)
+
+    @update_model_metadata_field.mapping.delete
```
Comment (matthewelwell) [inline api/metadata/mixins.py]: Ooh, what does this do?

File: api/metadata/models.py (hunk 5 above)
Comment (matthewelwell) [reply api/metadata/models.py:115]: Is there any way around it? Can the reverse relation be created based on related fields for example? Something like `model_field__content_type`. Feels weird that we're having to define the same data twice.

File: api/metadata/permissions.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class MetadataFieldPermissions(IsAuthenticated):
+    def has_permission(self, request, view):
+        if not super().has_permission(request, view):
+            return False
+
+        organisation = request.data.get("organisation")
+        if view.action == "create" and request.user.belongs_to(organisation):
+            return request.user.is_organisation_admin(request.data.get("organisation"))
```
Comment (matthewelwell) [inline api/metadata/permissions.py]: I think you can replace `request.data.get("organisation")` with just `organisation` since you've already stored it in a variable above. 

Also, it might be worth creating an organisation level permission for this (although perhaps that's something that we can do in a separate PR).

File: api/metadata/serializers.py [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+        model = MetadataModelField
+        fields = ("id", "field", "is_required")
+
+
+class FieldDataField(serializers.Field):
+    """
+    Custom field to handle the different types of data supported by the metadata field
+    """
+
+    def to_internal_value(self, data):
```
Comment (matthewelwell) [inline api/metadata/serializers.py]: Do we need to do this? Shouldn't we just get the client to send up the type and we can always just store the data as a string?

File: api/metadata/serializers.py [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+            raise serializers.ValidationError(
+                f"Value string is too long. Must be less than {FIELD_VALUE_MAX_LENGTH} character"
+            )
+        return data
+
+    def to_representation(self, value):
+        return value
+
+
+class MetadataListSerializer(serializers.ListSerializer):
```
Comment (matthewelwell) [inline api/metadata/serializers.py]: Should we update the name of this serializer since we're (seemingly) using it for more than just listing.

File: api/metadata/serializers.py [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+    def save(self, **kwargs):
```
Comment (matthewelwell) [inline api/metadata/serializers.py]: I'm a little confused what we're doing here - is this some kind of bulk create functionality?

File: api/metadata/serializers.py [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+        fields = ("model_field", "field_value")
+        list_serializer_class = MetadataListSerializer
+
+    def validate(self, data):
+        if type(data["field_value"]).__name__ != data["model_field"].field.type:
+            raise serializers.ValidationError("Invalid type")
+        return data
+
+    def to_representation(self, value):
+        # Convert field_value to its appropriate type(from string)
```
Comment (matthewelwell) [inline api/metadata/serializers.py]: Why though?

File: api/metadata/serializers.py [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+                raise serializers.ValidationError(
+                    f"Missing required metadata field {metadata_field.field.name}"
+                )
+
+    def save(self, **kwargs):
+        self.validate_required_metadata()
+
+        instance = super().save(**kwargs)
+
+        metadata = self.get_initial().get("metadata", None)
```
Comment (matthewelwell) [inline api/metadata/serializers.py]: Can we use a WritableNestedSerializer for this instead? or perhaps not because of all the GenericRelation stuff?

**Review by matthewelwell — COMMENTED** [review]
I've been through the logic (but not the test code yet) and added a few comments. Once we've nailed down all the questions on the logic then I can go through the tests but I feel like there might end up being a few changes to the tests based on the comments on the logic. 

I also noticed that there is an additional (probably erroneous) file that's been added at `api/output.txt` (annoyingly you can't leave comments on empty files in GH).

File: api/metadata/models.py (hunk 5 above)
Comment (gagantrivedi) [reply api/metadata/models.py:115]: Yeah, it does feel weird, but I don't think there is any other way to create a reverse relation

File: api/metadata/serializers.py (hunk 19 above)
Comment (gagantrivedi) [reply api/metadata/serializers.py]: Hmm, I think there was some reason (Ah no it was related to the old design) good shout

File: api/metadata/serializers.py (hunk 15 above)
Comment (gagantrivedi) [reply api/metadata/serializers.py]: But we already have the expected type for this, right? This is used to verify its type

File: api/import_export/export.py (hunk 8 above)
Comment (gagantrivedi) [reply api/import_export/export.py:104]: You mean with MetadataModelField ?

File: api/metadata/mixins.py (hunk 12 above)
Comment (gagantrivedi) [reply api/metadata/mixins.py]: I don't think that helps much, though, right? We still need to perform these checks, right?

File: api/metadata/mixins.py (hunk 13 above)
Comment (gagantrivedi) [reply api/metadata/mixins.py]: It basically provides the same `url_path` as the update method

File: api/metadata/serializers.py (hunk 18 above)
Comment (gagantrivedi) [reply api/metadata/serializers.py]: Let  front-end to do the type casting?

Comment (matthewelwell) [reply api/metadata/serializers.py]: Yep, I think for any new fields that can be multiple types, we should not do any type casting on the BE - it makes things very difficult! The FE can then use the string value and the returned type value to display it how they want.

File: api/import_export/export.py (hunk 8 above)
Comment (matthewelwell) [reply api/import_export/export.py:104]: Yes, sorry, I think I put the comment in the wrong place.

File: api/metadata/mixins.py (hunk 12 above)
Comment (matthewelwell) [reply api/metadata/mixins.py]: Well, it just means we don't have to dive into the request body to do permission checks which for some reason feels slightly wrong to me (although I'm aware we already have other scenarios that we do do this.

File: api/metadata/mixins.py (hunk 13 above)
Comment (matthewelwell) [reply api/metadata/mixins.py]: Ah, ok, so it's basically just shorthand for: 

```python
@action(
        detail=False,
        methods=["DELETE"],
        url_path=r"metadata-field/(?P<metadata_pk>\d+)",
    )
```

File: api/metadata/serializers.py (hunk 15 above)
Comment (matthewelwell) [reply api/metadata/serializers.py]: Hmm... I think we should just get the FE to send a string (so we don't have to hack the API docs) and then do something more like: 

```python
casting_functions = {"int": int, "str": str, ...}

try:
    casting_function = casting_functions.get(field_type)  # we might need to use context to get the field type here? 
    casting_function(data)
except ValueError:
    raise ValidationError(...)

if len(str(data)) > FIELD_VALUE_MAX_LENGTH:
    raise ValidationError(...)

return data
```

File: api/import_export/export.py (hunk 8 above)
Comment (gagantrivedi) [reply api/import_export/export.py:104]: It does have UUID, though?

Comment (matthewelwell) [reply api/import_export/export.py:104]: But it's not a traditional foreign key, right? So I'm not sure how it will link to the relevant entities when we export it (I can't quite remember how the export works tbh!)

File: api/metadata/mixins.py (hunk 9 above)
Comment (matthewelwell) [reply api/metadata/mixins.py]: Did this get removed instead?

Comment (gagantrivedi) [reply api/metadata/mixins.py]: Yep

File: api/metadata/mixins.py (hunk 10 above)
Comment (matthewelwell) [reply api/metadata/mixins.py]: Nice, just saw that we were able to use `Meta.model` 👌

File: api/metadata/serializers.py [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+                "str": str,
+                "bool": str_to_bool,
+            }[expected_type]
+            casting_function(data["field_value"])
+        except ValueError:
+            raise serializers.ValidationError(
+                f"Invalid data type. Must be the string representation of {expected_type}"
+            )
+
+        if expected_type == str and len(data["field_value"]) > FIELD_VALUE_MAX_LENGTH:
```
Comment (matthewelwell) [inline api/metadata/serializers.py]: I think this should be `"str"` ?

File: api/metadata/serializers.py [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+    def validate(self, data):
+        data = super().validate(data)
+        expected_type = data["model_field"].field.type
+        try:
+            casting_function = {
+                "int": int,
+                "float": float,
+                "str": str,
+                "bool": str_to_bool,
+            }[expected_type]
```
Comment (matthewelwell) [inline api/metadata/serializers.py]: Should we add a separate 'except KeyError` ? I guess the type can't be anything other than one of these but might be best to catch it just in case?

File: api/metadata/views.py [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+
+@method_decorator(
+    name="list",
+    decorator=swagger_auto_schema(
+        manual_parameters=[
+            openapi.Parameter(
+                "organisation",
+                openapi.IN_QUERY,
+                "ID of the organisation to filter by.",
+                required=False,
```
Comment (matthewelwell) [inline api/metadata/views.py]: I think we should probably make this required? I can't see any reason why we'd want to show fields outside of the context of an organisation?

File: api/metadata/permissions.py [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+from metadata.models import MetadataField
+
+
+class MetadataFieldPermissions(IsAuthenticated):
+    def has_permission(self, request, view):
+        if not super().has_permission(request, view):
+            return False
+
+        organisation = request.data.get("organisation")
+        if view.action == "create" and request.user.belongs_to(organisation):
```
Comment (matthewelwell) [inline api/metadata/permissions.py]: I'm sure why we need the `request.user.belongs_to(organisation)` clause? From the looks of it, that's covered by `request.user.is_organisation_admin(organisation)`

File: api/metadata/permissions.py [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+
+        # list is handled by the view
+        if view.action == "list":
+            return True
+
+        # move on to object specific permissions
+        return view.detail
+
+    def has_object_permission(self, request, view, obj):
+        if view.action in ("update", "destroy") and request.user.is_organisation_admin(
```
Comment (matthewelwell) [inline api/metadata/permissions.py]: What about `partial_update` and `retrieve` ? I think the ViewSet is a ModelViewSet so these actions will exist.

File: api/metadata/permissions.py [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+
+        return False
+
+
+class MetadataModelFieldPermissions(IsAuthenticated):
+    def has_permission(self, request, view):
+        if not super().has_permission(request, view):
+            return False
+
+        organisation_pk = int(view.kwargs.get("organisation_pk"))
```
Comment (matthewelwell) [inline api/metadata/permissions.py]: If the FE accidentally tries to call `/organisations/undefined/metadata-...` this will throw a hard exception, right? Obviously we'd expect it to error in that case but it would be nice if it failed more gracefully than raising a `TypeError`.

File: api/metadata/permissions.py [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+
+        if request.user.belongs_to(organisation_pk):
+            if view.action in ["list", "get_supported_content_types"] or view.detail:
+                return True
+
+            if view.action == "create":
+                field = MetadataField.objects.get(id=request.data.get("field"))
```
Comment (matthewelwell) [inline api/metadata/permissions.py]: Similar to above, this could throw `MetadataField.DoesNotExist`, right?

File: api/metadata/permissions.py [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+
+                return (
+                    request.user.is_organisation_admin(organisation_pk)
+                    and organisation_pk == field.organisation.id
+                )
+
+        return False
+
+    def has_object_permission(self, request, view, obj):
+        if view.action in ("update", "destroy") and request.user.is_organisation_admin(
```
Comment (matthewelwell) [inline api/metadata/permissions.py]: As above, what about the other actions?

File: api/metadata/serializers.py [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+        return data
+
+
+class MetadataSerializerMixin:
+    def validate_required_metadata(self, data):
+        metadata = data.get("metadata", [])
+
+        content_type = ContentType.objects.get_for_model(self.Meta.model)
+
+        required_metadata_fields = MetadataModelField.objects.filter(
```
Comment (matthewelwell) [inline api/metadata/serializers.py]: Shouldn't we be filtering this by organisation too?

File: api/metadata/tests/conftest.py [hunk 29]
```diff
@@ -0,0 +1,23 @@
+import pytest
```
Comment (matthewelwell) [inline api/metadata/tests/conftest.py]: Using this line to comment as it's the first line in a file (and Github still doesn't allow commenting on files as a whole!) 

Unless there's a specific reason not to, I think these test files should all be in `api/tests/metadata/...`?

File: api/tests/unit/import_export/test_unit_import_export_export.py [hunk 30]
```diff
@@ -125,6 +130,50 @@ def test_export_environments(project):
...
+    metadata_field = MetadataField.objects.create(
+        name="test_field", type="int", organisation=organisation
+    )
+
+    environment_metadata_field = MetadataModelField.objects.create(
+        field=metadata_field, content_type=environment_type
+    )
+
+    metadata = Metadata.objects.create(
+        object_id=environment.id,
```
Comment (matthewelwell) [inline api/tests/unit/import_export/test_unit_import_export_export.py:154]: One thing I'm keen to check here is whether this link to the environment will still work post import? Or if we could see some unexpected results here?

File: api/metadata/serializers.py (hunk 21 above)
Comment (gagantrivedi) [reply api/metadata/serializers.py]: Getting key Error here will mean that something terrible has happened(with the db?) and it should fail as loud as possible in that case, no?

Comment (matthewelwell) [reply api/metadata/serializers.py]: Yes, you're probably right. Let's leave it as is.

File: api/metadata/serializers.py [hunk 31]
```diff
@@ -0,0 +1,90 @@
...
+            )
+
+        if expected_type == "str" and len(data["field_value"]) > FIELD_VALUE_MAX_LENGTH:
+            raise serializers.ValidationError(
+                f"Value string is too long. Must be less than {FIELD_VALUE_MAX_LENGTH} character"
+            )
+        return data
+
+
+class MetadataSerializerMixin:
```
Comment (gagantrivedi) [inline api/metadata/serializers.py]: I am not sure if we should still call it a mixin. Thoughts?

File: api/metadata/serializers.py (hunk 28 above)
Comment (gagantrivedi) [reply api/metadata/serializers.py]: Yep, good catch

File: api/metadata/models.py [hunk 32]
```diff
@@ -0,0 +1,49 @@
...
+    organisation = models.ForeignKey(Organisation, on_delete=models.CASCADE)
+
+    class Meta:
+        unique_together = ("name", "organisation")
+
+
+class MetadataModelField(AbstractBaseExportableModel):
+    field = models.ForeignKey(MetadataField, on_delete=models.CASCADE)
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    is_required = models.BooleanField(default=False)
```
Comment (matthewelwell) [inline api/metadata/models.py]: Just chatted to @dabeeeenster and @kyle-ssg about this. Ben made a good point that users may not want metadata to be required across the entire organisation. So, for example, on features, you may not want to make JIRA ticket number required across all the projects in your organisation. We need to think about a way to make things 'required with respect to'. 

Can you have a think about this @gagantrivedi ?

File: api/metadata/models.py [hunk 33]
```diff
@@ -0,0 +1,49 @@
...
+from django.db import models
+
+from organisations.models import Organisation
+
+from .fields import GenericObjectID
+
+FIELD_VALUE_MAX_LENGTH = 2000
+
+
+class FieldType(models.TextChoices):
```
Comment (matthewelwell) [inline api/metadata/models.py:28]: We probably want to add more data types here, and make them less closely coupled to python data types. So for example, we probably want to split STRING into SINGLE_LINE_STRING and MULTI_LINE_STRING (primarily for the UI's benefit). 

I haven't looked through the rest of the PR to see how valid this might be but I'm imagining we would define custom validation functions against the field types.

Comment (dabeeeenster) [reply api/metadata/models.py:28]: Also we should have a `URL` data type

File: api/metadata/models.py (hunk 32 above)
Comment (gagantrivedi) [reply api/metadata/models.py]: Yeah, good point

File: api/import_export/export.py (hunk 8 above)
Comment (matthewelwell) [reply api/import_export/export.py:104]: I'm guessing this is resolved by the code in `json_serializers_with_metadata_support` ?

File: api/metadata/fields.py [hunk 34]
```diff
@@ -0,0 +1,29 @@
+from django.core import exceptions
+from django.db import models
+
+
+class GenericObjectID(models.PositiveIntegerField):
```
Comment (matthewelwell) [inline api/metadata/fields.py:5]: I might need you to run through this in a bit more detail with me @gagantrivedi.

File: api/metadata/models.py [hunk 35]
```diff
@@ -0,0 +1,126 @@
...
+from organisations.models import Organisation
+from projects.models import Project
+
+from .fields import GenericObjectID
+
+FIELD_VALUE_MAX_LENGTH = 2000
+
+METADATA_SUPPORTED_MODELS = ["environment"]
+
+# A map of model name to a function that takes the object id and returns the organisation_id
```
Comment (matthewelwell) [inline api/metadata/models.py:17]: I'm not sure why we have two levels here? What relevance does the `"environment"` top level key have?

File: api/metadata/models.py [hunk 36]
```diff
@@ -0,0 +1,126 @@
...
+        ).organisation.id,
+    }
+}
+
+
+class FieldType(models.TextChoices):
+    INTEGER = "int"
+    STRING = "str"
+    BOOLEAN = "bool"
+    FLOAT = "float"
```
Comment (matthewelwell) [inline api/metadata/models.py]: We should get rid of float I think

File: api/metadata/models.py [hunk 37]
```diff
@@ -0,0 +1,126 @@
...
+        "project": lambda project_id: Project.objects.get(
+            id=project_id
+        ).organisation.id,
+    }
+}
+
+
+class FieldType(models.TextChoices):
+    INTEGER = "int"
+    STRING = "str"
```
Comment (matthewelwell) [inline api/metadata/models.py:30]: We should split `STRING` into `MULTILINE_STRING` and `SINGLE_LINE_STRING`. Or perhaps just `STRING` and `MULTILINE_STRING` would be sufficient?

File: api/metadata/models.py [hunk 38]
```diff
@@ -0,0 +1,126 @@
...
+
+    name = models.CharField(max_length=255)
+    type = models.CharField(
+        max_length=255, choices=FieldType.choices, default=FieldType.STRING
+    )
+    description = models.TextField(blank=True, null=True)
+    organisation = models.ForeignKey(Organisation, on_delete=models.CASCADE)
+
+    def is_field_value_valid(self, field_value: str) -> bool:
+        if len(field_value) > FIELD_VALUE_MAX_LENGTH:
```
Comment (matthewelwell) [inline api/metadata/models.py:47]: If we have multiline strings, we may need to have different max length values depending on the type?

File: api/metadata/models.py [hunk 39]
```diff
@@ -0,0 +1,126 @@
...
+
+    def validate_url(self, field_value: str):
+        try:
+            result = urlparse(field_value)
+            return all([result.scheme, result.netloc])
+        except ValueError:
+            return False
+
+    def validate_str(self, field_value: str):
+        return True
```
Comment (matthewelwell) [inline api/metadata/models.py:71]: Do we need to make sure we convert the value to `str` here (and in the other methods?) ?

File: api/metadata/views.py [hunk 40]
```diff
@@ -0,0 +1,124 @@
...
+        ]
+    ),
+)
+class MetadataFieldViewSet(viewsets.ModelViewSet):
+    permission_classes = [MetadataFieldPermissions]
+    serializer_class = MetadataFieldSerializer
+
+    def get_queryset(self):
+        queryset = MetadataField.objects.filter(
+            organisation__in=self.request.user.organisations.all()
```
Comment (matthewelwell) [inline api/metadata/views.py]: Couldn't this be: `organisation__user=self.request.user` ?

File: api/metadata/permissions.py [hunk 41]
```diff
@@ -0,0 +1,77 @@
...
+from organisations.models import Organisation
+
+
+class MetadataFieldPermissions(IsAuthenticated):
+    def has_permission(self, request, view):
+        if not super().has_permission(request, view):
+            return False
+
+        # list is handled by the view
+        if view.action == "list" or view.detail:
```
Comment (matthewelwell) [inline api/metadata/permissions.py:15]: Perhaps not super important but I think we _might_ want to implement a permission check for the `list` action: it would be better for the user to see a 'you do not have permission' vs an empty list, right?

File: api/metadata/permissions.py [hunk 42]
```diff
@@ -0,0 +1,77 @@
...
+            return True
+
+        if view.action == "create":
+            with suppress(Organisation.DoesNotExist):
+                organisation_id = request.data.get("organisation")
```
Comment (matthewelwell) [inline api/metadata/permissions.py:20]: Shouldn't we get this from `self.kwargs` rather than the request data? Can we remove `organisation` from the request data?

File: api/metadata/serializers.py [hunk 43]
```diff
@@ -0,0 +1,134 @@
...
+                    data["content_type"].model
+                ][requirement["content_type"].model]
+            except KeyError:
+                raise serializers.ValidationError(
+                    "Invalid requirement for model {}".format(
+                        data["content_type"].model
+                    )
+                )
+
+                if (
```
Comment (matthewelwell) [inline api/metadata/serializers.py]: I think this code is unreachable? It should be one level higher, right?

File: api/metadata/models.py (hunk 35 above)
Comment (matthewelwell) [reply api/metadata/models.py:17]: I guess this is helping somehow when we want to say that 'metadata field x is required for this project / organisation' ?

File: api/metadata/models.py [hunk 44]
```diff
@@ -0,0 +1,126 @@
...
+
+METADATA_SUPPORTED_MODELS = ["environment"]
+
+# A map of model name to a function that takes the object id and returns the organisation_id
+SUPPORTED_REQUIREMENTS_MAPPING = {
+    "environment": {
+        "organisation": lambda org_id: org_id,
+        "project": lambda project_id: Project.objects.get(
+            id=project_id
+        ).organisation.id,
```
Comment (matthewelwell) [inline api/metadata/models.py]:
```suggestion
        ).organisation_id,
```

File: api/metadata/serializers.py (hunk 31 above)
Comment (matthewelwell) [reply api/metadata/serializers.py]: Why not?

File: api/metadata/serializers.py [hunk 45]
```diff
@@ -0,0 +1,134 @@
...
+        raise NotImplementedError()
+
+    def get_required_for_object(
+        self, requirement: MetadataModelFieldRequirement, data: dict
+    ) -> Model:
+        model_name = requirement.content_type.model
+        try:
+            return getattr(self, f"get_{model_name}_from_validated_data")(data)
+        except AttributeError:
+            raise ValueError(f"`get_{model_name}_from_validated_data` does not exists")
```
Comment (matthewelwell) [inline api/metadata/serializers.py]:
```suggestion
            raise ValueError(f"`get_{model_name}_from_validated_data` method does not exist")
```

Comment (matthewelwell) [reply api/metadata/serializers.py:102]: Should this be a `NotImplementedError` ? 🤔

File: api/metadata/views.py [hunk 46]
```diff
@@ -0,0 +1,124 @@
...
+            if organisation_id is None:
+                raise ValidationError("organisation parameter is required")
+            queryset = queryset.filter(organisation__id=organisation_id)
+
+        return queryset
+
+
+@method_decorator(
+    name="list",
+    decorator=swagger_auto_schema(
```
Comment (matthewelwell) [inline api/metadata/views.py]: Couldn't we use a query_serializer for this instead?

File: api/metadata/views.py [hunk 47]
```diff
@@ -0,0 +1,124 @@
...
+        ]
+    ),
+)
+class MetaDataModelFieldViewSet(viewsets.ModelViewSet):
+    permission_classes = [MetadataModelFieldPermissions]
+    serializer_class = MetaDataModelFieldSerializer
+
+    def get_queryset(self):
+        queryset = MetadataModelField.objects.filter(
+            field__organisation_id=self.kwargs.get("organisation_pk")
```
Comment (matthewelwell) [inline api/metadata/views.py:61]: Are we doing a permission check for this somewhere else?

File: api/metadata/serializers.py (hunk 31 above)
Comment (matthewelwell) [reply api/metadata/serializers.py]: Yep, ok, perhaps this should inherit from `serializers.BaseSerializer` and be named something like: `SerializerWithMetadata`

File: api/metadata/models.py (hunk 39 above)
Comment (gagantrivedi) [reply api/metadata/models.py:71]: That's not needed since it's a text field and anything can be converted to string, right?

File: api/metadata/models.py (hunk 38 above)
Comment (gagantrivedi) [reply api/metadata/models.py:47]: I don't think people want to go above 2K?

File: api/metadata/permissions.py (hunk 41 above)
Comment (gagantrivedi) [reply api/metadata/permissions.py:15]: As discussed - this view is not nested under organisation

File: api/metadata/models.py (hunk 38 above)
Comment (matthewelwell) [reply api/metadata/models.py:47]: Perhaps not, I guess we can deal with this later if we need to.

**Review by matthewelwell — APPROVED** [review]
I've approved this now but let's set up a meeting next week (preferably with @dabeeeenster and @kyle-ssg) to review the functionality and make sure that it fulfils the requirements correctly.
